File: docs/absorbed-from-phenotype-gfx/unity/terrain/src/TerrainLod.cs
using System;

namespace Phenotype.Terrain
{
    /// <summary>
    /// Manages level-of-detail selection for terrain chunks based on camera distance.
    /// Coordinates mesh swap-ins to keep draw calls bounded at wide zoom levels.
    /// </summary>
    /// <remarks>
    /// Default thresholds are <c>50 / 150 / 400</c> world units and default resolutions are
    /// <c>64 / 32 / 16</c> quads per side. Override these values to tune for your scene scale.
    /// </remarks>
    /// <example>
    /// <code>
    /// var lod = new TerrainLod
    /// {
    ///     NearDistance = 30f,
    ///     MidDistance = 120f,
    ///     CullDistance = 500f,
    ///     NearResolution = 128,
    ///     MidResolution = 64,
    ///     FarResolution = 32,
    /// };
    ///
    /// int res = lod.SelectResolution(80f); // 64
    /// </code>
    /// </example>
    public class TerrainLod : LodBase
    {
        /// <summary>
        /// Distance to <see cref="LodTier.Mid"/> transition. Default: 50.
        /// </summary>
        /// <value>Must be less than <see cref="MidDistance"/>.</value>
        /// <example>
        /// <code>
        /// var lod = new TerrainLod();
        /// lod.NearDistance = 40f;
        /// </code>
        /// </example>
        public override float NearDistance { get; set; } = 50f;

        /// <summary>
        /// Distance to <see cref="LodTier.Far"/> transition. Default: 150.
        /// </summary>
        /// <value>Must be greater than <see cref="NearDistance"/> and less than <see cref="CullDistance"/>.</value>
        /// <example>
        /// <code>
        /// var lod = new TerrainLod();
        /// lod.MidDistance = 200f;
        /// </code>
        /// </example>
        public override float MidDistance { get; set; } = 150f;

        /// <summary>
        /// Distance to <see cref="LodTier.Culled"/> transition. Default: 400.
        /// </summary>
        /// <value>Must be greater than <see cref="MidDistance"/>.</value>
        /// <example>
        /// <code>
        /// var lod = new TerrainLod();
        /// lod.CullDistance = 1000f;
        /// </code>
        /// </example>
        public override float CullDistance { get; set; } = 400f;

        /// <summary>
        /// Grid resolution used for the <see cref="LodTier.Near"/> tier. Default: 64.
        /// </summary>
        /// <value>Must be a positive integer.</value>
        /// <example>
        /// <code>
        /// var lod = new TerrainLod();
        /// lod.NearResolution = 128;
        /// </code>
        /// </example>
        public int NearResolution { get; set; } = 64;

        /// <summary>
        /// Grid resolution used for the <see cref="LodTier.Mid"/> tier. Default: 32.
        /// </summary>
        /// <value>Must be a positive integer. Typically half of <see cref="NearResolution"/>.</value>
        /// <example>
        /// <code>
        /// var lod = new TerrainLod();
        /// lod.MidResolution = 48;
        /// </code>
        /// </example>
        public int MidResolution { get; set; } = 32;

        /// <summary>
        /// Grid resolution used for the <see cref="LodTier.Far"/> tier. Default: 16.
        /// </summary>
        /// <value>Must be a positive integer. Typically half of <see cref="MidResolution"/>.</value>
        /// <example>
        /// <code>
        /// var lod = new TerrainLod();
        /// lod.FarResolution = 8;
        /// </code>
        /// </example>
        public int FarResolution { get; set; } = 16;

        /// <summary>
        /// Returns the grid resolution appropriate for the given camera distance.
        /// Returns 0 when the mesh should be culled.
        /// </summary>
        /// <param name="distance">Camera-to-mesh distance in world units. Must be >= 0.</param>
        /// <returns>
        /// The grid resolution for the selected <see cref="LodTier"/>:
        /// <list type="table">
        /// <listheader><term>Tier</term><description>Resolution</description></listheader>
        /// <item><term><see cref="LodTier.Near"/></term><description><see cref="NearResolution"/></description></item>
        /// <item><term><see cref="LodTier.Mid"/></term><description><see cref="MidResolution"/></description></item>
        /// <item><term><see cref="LodTier.Far"/></term><description><see cref="FarResolution"/></description></item>
        /// <item><term><see cref="LodTier.Culled"/></term><description>0</description></item>
        /// </list>
        /// </returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when distance is negative.</exception>
        /// <example>
        /// <code>
        /// var lod = new TerrainLod();
        /// int nearRes = lod.SelectResolution(25f);  // 64
        /// int midRes = lod.SelectResolution(100f);   // 32
        /// int farRes = lod.SelectResolution(300f);   // 16
        /// int culled = lod.SelectResolution(500f);   // 0
        /// </code>
        /// </example>
        public int SelectResolution(float distance)
        {
            return SelectTier(distance) switch
            {
                LodTier.Near => NearResolution,
                LodTier.Mid => MidResolution,
                LodTier.Far => FarResolution,
                LodTier.Culled => 0,
                _ => 0,
            };
        }
    }
}

File: docs/absorbed-from-phenotype-gfx/src/terrain/terrain_lod.rs
//! Concrete `TerrainLod` impl. Ported from C# `TerrainLod.cs`.
//!
//! Default thresholds: 50 / 150 / 400 world units. Default resolutions:
//! 64 / 32 / 16 quads per side.

use serde::{Deserialize, Serialize};

use crate::terrain::error::TerrainResult;
use crate::terrain::lod::{LodBase, LodTier};

/// Default LOD config for terrain chunks.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct TerrainLod {
    near: f32,
    mid: f32,
    cull: f32,
    near_resolution: i32,
    mid_resolution: i32,
    far_resolution: i32,
}

impl Default for TerrainLod {
    fn default() -> Self {
        Self {
            near: 50.0,
            mid: 150.0,
            cull: 400.0,
            near_resolution: 64,
            mid_resolution: 32,
            far_resolution: 16,
        }
    }
}

impl TerrainLod {
    /// Create with default thresholds and resolutions.
    pub fn new() -> Self { Self::default() }

    /// Grid resolution for the `Near` tier. Must be `> 0`.
    pub fn near_resolution(&self) -> i32 { self.near_resolution }
    /// Set the near resolution. Caller is responsible for `> 0`.
    pub fn set_near_resolution(&mut self, v: i32) { self.near_resolution = v; }

    /// Grid resolution for the `Mid` tier.
    pub fn mid_resolution(&self) -> i32 { self.mid_resolution }
    /// Set the mid resolution.
    pub fn set_mid_resolution(&mut self, v: i32) { self.mid_resolution = v; }

    /// Grid resolution for the `Far` tier.
    pub fn far_resolution(&self) -> i32 { self.far_resolution }
    /// Set the far resolution.
    pub fn set_far_resolution(&mut self, v: i32) { self.far_resolution = v; }

    /// Returns the grid resolution for the given distance; `0` when culled.
    pub fn select_resolution(&self, distance: f32) -> TerrainResult<i32> {
        Ok(match self.select_tier(distance)? {
            LodTier::Near => self.near_resolution,
            LodTier::Mid => self.mid_resolution,
            LodTier::Far => self.far_resolution,
            LodTier::Culled => 0,
        })
    }
}

impl LodBase for TerrainLod {
    fn near_distance(&self) -> f32 { self.near }
    fn set_near_distance(&mut self, v: f32) { self.near = v; }
    fn mid_distance(&self) -> f32 { self.mid }
    fn set_mid_distance(&mut self, v: f32) { self.mid = v; }
    fn cull_distance(&self) -> f32 { self.cull }
    fn set_cull_distance(&mut self, v: f32) { self.cull = v; }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_thresholds_are_expected() {
        let lod = TerrainLod::new();
        assert_eq!(lod.near_distance(), 50.0);
        assert_eq!(lod.mid_distance(), 150.0);
        assert_eq!(lod.cull_distance(), 400.0);
    }

    #[test]
    fn default_resolutions_are_expected() {
        let lod = TerrainLod::new();
        assert_eq!(lod.near_resolution(), 64);
        assert_eq!(lod.mid_resolution(), 32);
        assert_eq!(lod.far_resolution(), 16);
    }

    #[test]
    fn select_resolution_near_distance_returns_near_resolution() {
        let lod = TerrainLod::new();
        assert_eq!(lod.select_resolution(25.0).unwrap(), 64);
    }

    #[test]
    fn select_resolution_mid_distance_returns_mid_resolution() {
        let lod = TerrainLod::new();
        assert_eq!(lod.select_resolution(100.0).unwrap(), 32);
    }

    #[test]
    fn select_resolution_far_distance_returns_far_resolution() {
        let lod = TerrainLod::new();
        assert_eq!(lod.select_resolution(250.0).unwrap(), 16);
    }

    #[test]
    fn select_resolution_culled_returns_zero() {
        let lod = TerrainLod::new();
        assert_eq!(lod.select_resolution(500.0).unwrap(), 0);
    }

    #[test]
    fn validate_thresholds_default_does_not_throw() {
        let lod = TerrainLod::new();
        assert!(lod.validate_thresholds().is_ok());
    }
}
